include/istio/control/tcp/check_data.h

```diff
@@ -34,6 +34,10 @@ class CheckData {
   // If SSL is used, get peer or local certificate SAN URI.
   virtual bool GetPrincipal(bool peer, std::string* user) const = 0;
 
+  // Get source namespace from the principal.
+  virtual bool GetSourceNamespace(const std::string& principal,
```

Comment:
Are we always going to extract namespace from principal?  If so,  we dont need this function, top level (whoever is calling this object) can just generate it from principal. 

It is also odd, that I did not see this function for http/check_data.   I think it is extracted from dynamic metadata generated by authn.    Are we sure authn filter is always there?

Reply:
I called GetSourceNamespace() in utility.cc directly in Authentication::SaveAuthAttributesToStruct(), so that's why it's not in http/check_data.

I added it to tcp/check_data mainly because It seems I cannot make `istio/control/tcp/attributes_builder.cc` to depend on `envoy/utils/utils.cc`. Do you want me to create the `istio/utils.utils.cc` to include `GetSourceNamespace()` and use it directly from http and tcp?

For your last question, I think if authn filter is not there, this should only happen when mTLS is disabled, in such case, we could still let the Mixer to generate the source.namespace attribute.

Reply:
Yes, I think it will be better if you 
 move extracting namespace from principal into istio/utils/utils.cc  so that tcp/attribute_builder.cc can use it and http/authn filter can use it too.  so not need to change tcp/check_data.h

Reply:
@qiwzhang raises a good point. authn_filter currently not always there, even for http. Basically, we don't add it when no authN needed. It's just a small optimization, as there is nothing to do/extract when auth is disabled. However, as we increase the dependency between filters, I think it's better to keep it there always, so it's more consistent in the way we build authenticated attributes.

Reply:
Done